accepted/exception-handling.md

```diff
@@ -0,0 +1,240 @@
+possibly cheap enough that is acceptable in the interim.
+
+Concretely, we propose to implement non-zero cost exceptions by using
+flags (e.g. the overflow flag) to determine whether a call returned
+through a throw or an ordinary return instruction. Put into code:
+
+```
+call 0x12345678 # exceptional return sets the flag
+jo $exception_branch
+```
+
+Before returning normally any function must clear the flag, e.g.
+
+```
+test al, al
+ret
+```
+
+We reckon this approach is relatively low overhead, and it something
+we can confidently implement correctly more quickly than the side
+table or DWARF unwinder strategies. Adopting this strategy would allow
+us to focus on the challenges of retrofitting exception support onto
+CLIF and extending the Wasmtime public API first. After the fact, we
+can focus the effort on getting all the nitty-gritty runtime bits for
+interpreting unwind info correct.
+
+## Unwinding across instances
+[unwinding-instances]: #unwinding-across-instances
+
+The calling convention strategy lets us aptly support unwinding across
+instances, i.e. scenarios such as
+
+```
+instance_A -> instance_B -> raise exception E
+```
+
+where instance `A` has installed an exception handler for `E` and
+calls into instance `B` which raises the exception `E`.
+
+To support unwinding across instances we plan to use a callee-saved
+register to hold (a pointer to) the exception `E`. This way instance
+`A` can readily get hold of the exception data.
+
+## Unwinding across host frames
+[unwinding-hosts]: #unwinding-across-host-frames
+
+As stated in the [non-requirements](:non-requirements) section we do
+not plan to support unwinding across host frames. However, to
+gracefully handle exceptions that attempt to cross wasm-host boundary
+we propose to have our host-to-wasm trampoline catch any Wasm-thrown
+exceptions and return them via a three-way result type.
+
+Similarly, we will do not plan to supporting raising Wasm exceptions
+directly in host code. Instead, we envisage a host function supply an
+element of the exception type as an argument, which the wasm-to-host
+trampolines will reify as a genuine exception and raise inside the
+Wasm code.
+
+For the design of the three-way result type there are multiple options
+to consider:
+
+* A genuine three-parameter result type capturing the three ways in
+  which a Wasm program can terminate, e.g. `Result<Success, Exception,
+  Trap>`.
+* A nested result type, e.g. `Result<Result<Success, Exception>,
+  Trap>` to remain compatible with Rust's `?` operator.
+* Or continue to use `Result<Success, anyhow::Error>` and allow
+  callers to try to downcast the `anyhow::Error` to either a trap or
+  an exception as needed.
+
+# Open questions
+[open-questions]: #open-questions
+
+* We need to decide whether `catch` blocks in CLIF are allowed to use
+  values from dominating blocks.
+* How should we represent three-way results in the Wasmtime public
+  API?
```

Comment:
For this open question I think it'd be reasonable to take inspiration from the JS API for exception handling, notably we'd have exported `Tag` structures which the embedder could create or acquire from instances. Exceptions themselves would probably be modeled as something that can be converted to `anyhow::Error` and then host functions would return that error to indicate they want to throw an exception. Afterwards how this is represented internally from that point is just an implementation detail.